src/Datadog.NativeLoader/pal.cpp

```diff
@@ -0,0 +1,102 @@
+#include "pal.h"
+
+#if _WIN32
+#include <Windows.h>
+#else
+#include "dlfcn.h"
+#endif
+
+#include "logging.h"
+
+namespace datadog
+{
+namespace nativeloader
+{
+
+    void* LoadDynamicLibrary(std::string filePath)
+    {
+        Debug("LoadLibrary: ", filePath);
+
+#if _WIN32
+        return LoadLibrary(ToWSTRING(filePath).c_str());
+#else
+        void* dynLibPtr = dlopen(filePath.c_str(), RTLD_LOCAL | RTLD_LAZY);
+        if (dynLibPtr == nullptr)
+        {
+            char* errorMessage = dlerror();
+            Warn("Error loading dynamic library: ", errorMessage);
+        }
+        return dynLibPtr;
+#endif
+    }
+
+    void* GetExternalFunction(void* instance, std::string funcName)
+    {
+        Debug("GetExternalFunction: ", funcName);
+
+#if _WIN32
+        return (void*) GetProcAddress((HMODULE) instance, funcName.c_str());
+#else
+        void* dynFunc = dlsym(instance, funcName.c_str());
+        if (dynFunc == nullptr)
+        {
+            char* errorMessage = dlerror();
+            Warn("Error loading dynamic function: ", errorMessage);
+        }
+        return dynFunc;
+#endif
+    }
+
+    WSTRING GetEnvironmentValue(const WSTRING& name)
+    {
+        Debug("GetEnvironmentValue: ", name);
+
+#ifdef _WIN32
+        const size_t max_buf_size = 4096;
+        WSTRING buf(max_buf_size, 0);
+        auto len = GetEnvironmentVariable(name.data(), buf.data(), (DWORD)(buf.size()));
+        return Trim(buf.substr(0, len));
+#else
+        auto cstr = std::getenv(ToString(name).c_str());
+        if (cstr == nullptr)
+        {
+            return WStr("");
+        }
+        std::string str(cstr);
+        auto wstr = ToWSTRING(str);
+        return Trim(wstr);
+#endif
```

Comment:
why not having a windows specific implementation? why not having the same for all platforms and base it on the std ?

Reply:
Is based on the implementation we already have in the tracer, that's why I added it in the pal file. To be honest I don't know if it will make any difference...

Reply:
👍

Reply:
well apparently doesn't have the same internals either: https://github.com/curl/curl/issues/4774

Reply:
ok thx #todayIlearn

Reply:
From that issue:

```log
Hi all,
I've noticed that curl_getenv() uses getenv() on Windows[1]. However,
environment variables set with SetEnvironmentVariable() are not seen by
getenv() (although GetEnvironmentVariable() sees changes done by
putenv()), and since SetEnvironmentVariable() is preferable to putenv()
because the former is thread-safe, this can be an issue. At least it
affects thread-safe PHP builds[2].
Is there anything that prevents libcurl to use GetEnvironmentVariable()
instead of getenv()?
```